### PR TITLE
Reduce cmx sizes by sharing variable names

### PR DESCRIPTION
**Pull request by lpw25**
This PR implements the first of the three optimisations in:

https://github.com/ocaml/ocaml/pull/1343

In particular, it makes sure that we get good sharing for the variable names generated by flambda. It takes a slightly different approach than the previous patch. The earlier one used a variant type for variable names, this one keeps it as a string but declares all the internal name in a single location and removes unnecessary uses of `^`. It produces a slightly better reduction in size than this part of the original patch. It is still only ~5% improvement, but that's not to be sniffed at and there isn't much additional complexity.

### CODE REVIEW

**Comment by gasche** [comment]
It is not clear to me that using well-shared strings is nicer than using constructors; I would have thought the converse, given that constructors enforce the sharing (instead of respecting it by convention), can be pattern-matched over, and more clearly expose the structure of the data (for example, `Block_symbol_get of int` clearly exposes a family of names, which is nicer than using `"block_symbol_get_" ^ pos_str`). What are your reasons to prefer shared strings?

**Comment by lpw25** [comment]
My main reason is that I intend to replace all uses of `Ident.t` after the typedtree with `Variable.t`, and so I don't want `Variable.t` to be flambda-specific. However, it occurs to me that when I do that I could make an flambda-specific variable that is `= private Variable.t` and get the same benefits, whilst allowing me to enforce that the name strings are shared by requiring them to come from `Internal_variable_names`. I could even add a functor to `Variable` for creating these private types and use it to create specific instances for each IR in the compiler, which would be nice.

That would address the problem of enforcing sharing via typing rather than convention. As for the other points: I disagree that pattern matching would be useful, or that the structure is better exposed because we want `Block_symbol_get foo` to behave identically to `String ("block_symbol_get_" ^ (int_of_string foo))` otherwise the name is being used as something other than a name.

**Comment by lpw25** [comment]
Also, using a string representation improves the size of `Variable.t`s who's name comes from the source, which is the common case.

**Comment by lpw25** [comment]
Ok, I've changed it so that `Internal_variable_names` has a type `t = private string`, and `Variable.create` requires an `Internal_variable_name.t` instead of a string. That should enforce sharing.

File: middle_end/ref_to_variables.ml [hunk 1]
```diff
@@ -16,9 +16,6 @@
 
 [@@@ocaml.warning "+a-4-9-30-40-41-42"]
 
-let rename_var var =
-  Mutable_variable.create
-    (Variable.unique_name var)
   (* Variable.rename var *)
   (*   ~current_compilation_unit:(Compilation_unit.get_current_exn ()) *)
```
Comment (xclerc) [inline middle_end/ref_to_variables.ml:20]: These floating comments are probably a left-over after the deletion of `rename_var`.

**Comment by xclerc** [comment]
Looks good to me. Unless I am mistaken, the only observable
change is that some identifiers will be slightly different (*e.g.* will
lose their `_{copied,original}` suffix), but it should not be a problem.

**Comment by mshinwell** [comment]
I will vouch for @xclerc 's review on this one (@lpw25 has looked at this too, of course, after it was originally written).

I have fixed the conflicts and run tests (both with and without flambda) on my own machine, so merging.

**Pull request merged**